### PR TITLE
Advanced text

### LINKED ISSUE #33: Text shaping and font fallback
[`wgpu_glyph`] uses [`glyph-brush`], which in turn uses [`rusttype`]. While the current implementation is able to layout text quite nicely, it does not perform any [text shaping].

[Text shaping] with font fallback is a necessary feature for any serious GUI toolkit. It unlocks support to truly localize your application, supporting many different scripts.

The only available library that does a great job at shaping is [HarfBuzz], which is implemented in C++. [`skribo`] seems to be a nice [HarfBuzz] wrapper for Rust.

This feature will probably imply rewriting [`wgpu_glyph`] entirely, as caching will be more complicated and the API will probably need to ask for more data.

[`wgpu_glyph`]: https://github.com/hecrj/wgpu_glyph
[`glyph-brush`]: https://github.com/alexheretic/glyph-brush
[`rusttype`]: https://github.com/redox-os/rusttype
[text shaping]: https://en.wikipedia.org/wiki/Complex_text_layout
[HarfBuzz]: https://github.com/harfbuzz/harfbuzz
[`skribo`]: https://github.com/linebender/skribo

### LINKED ISSUE #41: Fix text caching in `iced_wgpu`
Since we added scrollables (see #35), `iced_wgpu` supports different layers of primitives.

However, [`wgpu_glyph`](https://docs.rs/wgpu_glyph/0.5.0/wgpu_glyph/) does not support drawing text without trimming the cache. In other words, whenever we render multiple layers, we lose text caching.

We will most likely need to contribute to [`glyph-brush`](https://github.com/alexheretic/glyph-brush) to tackle this.

### LINKED ISSUE #1071: Software renderer
This issue is to measure interest and demonstrate support for the development of a software renderer for Iced

As described in the [roadmap](https://github.com/iced-rs/iced/blob/93fec8d273ef8305e1c2456abe0c8ecd7a9d9407/ROADMAP.md#reuse-existing-2d-renderers), two choices may be considered:
- [ ] [`skia`](https://skia.org), existing issue: https://github.com/iced-rs/iced/issues/520
- [ ] [`piet`](https://github.com/linebender/piet)

### LINKED ISSUE #1199: Support choosing a `Renderer` at runtime
Currently, `iced` forces users to pick a `Renderer` implementation at compilation time via feature flags (`iced_wgpu`, `iced_glow`, etc.).

However, each renderer has its tradeoffs and GUI applications may want to ship different renderers in the same binary to improve compatibility, performance, and correctness.

`iced` should allow users to enable renderers via feature flags in a non-exclusive way. Then, users could set the priority for each renderer in the `Settings` of an `Application`. For instance:

```rust
Application::run(Settings {
    renderers: [Renderer::Wgpu, Renderer::Glow],
    ..Settings::default()
})
```

I believe that in order to achieve this we will need to create a couple of new subcrates:

- `iced_renderer` on top of `iced_wgpu` and `iced_glow` and implementing a `Renderer` that picks its internal backend at runtime.
- `iced_shell` on top of `iced_winit` and`iced_glutin` implementing a generic shell that spawns and manages windows for the chosen renderer.

Once we have these, it should be relatively straightforward to change the root crate and extend the `Settings` struct.

### PR DESCRIPTION
**Pull request by hecrj**
This PR brings all the improvements that have been introduced in the `advanced-text` branch. Specifically:

- #1697
- #1748
- #1811
- #1822 
- #1828

It's hard to overstate the importance of these changes! All of these features combined make one of the biggest changesets to land in `master` since the inception of the library!

These efforts should lay the foundations for acquiring maturity in the long run (multi-line text input, accessibility, rendering parallelization, and more!). For more details, check out [the latest roadmap diagram](https://whimsical.com/roadmap-iced-7vhq6R35Lp3TmYH4WeYwLM).

Closes #33.
Fixes #41.
Closes #1071.
Closes #1199.

### CODE REVIEW

**Comment by GyulyVGC** [comment]
This is a complete revolution!
I'm so excited to see what iced will accomplish in the future 🚀

**Comment by gabrieldechichi** [comment]
Bug report: `image` seems broken on this branch.

Repro steps: Create a minimum iced app with a single image.

```rust
use iced::{Application, Settings};

fn main() -> iced::Result {
    app::App::run(iced::Settings {
        window: iced::window::Settings {
            size: (400, 300),
            position: iced::window::Position::Centered,
            ..Default::default()
        },
        ..Default::default()
    })
}

pub mod app {
    use iced::{widget::image, Sandbox};

    #[derive(Default)]
    pub struct App;

    #[derive(Debug, Clone, Copy)]
    pub enum Message {}

    impl Sandbox for App {
        type Message = Message;

        fn new() -> Self {
            App {}
        }

        fn title(&self) -> String {
            String::from("My Title")
        }

        fn update(&mut self, message: Self::Message) {}

        fn view(&self) -> iced::Element<'_, Self::Message> {
            image(image::Handle::from_path("resources/bg2.png"))
                .content_fit(iced::ContentFit::Fill)
                .into()
        }
    }
}
```

With `advanced-text`:

image

With iced 0.9.0:

image

**Comment by hecrj** [comment]
@gabrieldechichi Cannot reproduce.

**Comment by gabrieldechichi** [comment]
@hecrj @bungoboingo That's weird, I just had the same results in a different machine (same graphics cards though).

System specs:
- OS: Windows 11
- Graphics adapter: NVIDIA Geforce RTX 3060 Laptop GPU
- Graphics backend: Vulkan

I'm unsure how to check what iced renderer I'm using (sorry, I'm new to iced/rust), can you tell me how I can look this up?

I just a new project in the `examples` folder in a clone of this repo, checked out to this PR and pointing to the local `iced` source. These are the images I used to test (both have bugs):

[resources.zip](https://github.com/iced-rs/iced/files/11435227/resources.zip)

**Comment by gabrieldechichi** [comment]
Thanks @bungoboingo. So it looks like I'm running wgpu + Vulkan by default (sharing full logs in case anyone is interested).

I just tested, and the **images load correctly** with `tiny-skia`. So it could be an issue with wgpu + Vulkan + possibly my specific graphics card an drivers.

For context, the tour example also works for me, it's unclear if this is related to the specific image files, or the specific code I'm writing. If anyone is interested to test, this is a zip with everything (main.rs on my first post + resources). Assumes it is placed on `iced/examples` folder.

[image-test.zip](https://github.com/iced-rs/iced/files/11435633/image-test.zip)

Don't let this stall this PR if it's a non-repro though. Also let me know if this is not the correct place to raise this, and it's preferred to open issues.

<details>
<summary>Logs</summary>

```
[WARN  wgpu_hal::vulkan::instance] Unable to find layer: VK_LAYER_KHRONOS_validation
[INFO  wgpu_hal::vulkan::instance] Instance version: 0x4030e0
[INFO  wgpu_hal::vulkan::instance] Enabling debug utils
[INFO  wgpu_hal::vulkan::instance] Enabling device properties2
[INFO  wgpu_hal::gles::egl] Loading X11 library to get the current display
[INFO  wgpu_hal::gles::egl] EGL_MESA_platform_surfaceless not available. Using default platform
[INFO  wgpu_hal::gles::egl] Display vendor "Google Inc. (adapter LUID: 0000000000012047)", version (1, 4)
[WARN  wgpu_hal::gles::egl]     EGL surface: -srgb
[INFO  wgpu_hal::gles::egl]     Trying native-render
[WARN  wgpu_hal::gles::egl] No config found!
[INFO  wgpu_hal::gles::egl]     Trying presentation
[WARN  wgpu_hal::gles::egl] EGL says it can present to the window but not natively
[INFO  wgpu_hal::gles::egl]     EGL context: +debug KHR
[INFO  wgpu_hal::gles::egl]     EGL context: +robust access EXT
[INFO  wgpu_hal::gles::egl]     EGL context: +surfaceless
[INFO  iced_wgpu::window::compositor] Settings {
        present_mode: AutoVsync,
        internal_backend: Backends(
            VULKAN | GL | METAL | DX12 | DX11 | BROWSER_WEBGPU,
        ),
        default_font: Font {
            family: SansSerif,
            weight: Normal,
            stretch: Normal,
            monospaced: false,
        },
        default_text_size: 16.0,
        antialiasing: Some(
            MSAAx4,
        ),
    }
[INFO  wgpu_core::instance] Adapter Vulkan AdapterInfo { name: "NVIDIA GeForce RTX 3060 Laptop GPU", vendor: 4318, device: 9568, device_type: DiscreteGpu, driver: "NVIDIA", driver_info: "528.24", backend: Vulkan }
[INFO  wgpu_core::instance] Adapter Dx12 AdapterInfo { name: "NVIDIA GeForce RTX 3060 Laptop GPU", vendor: 4318, device: 9568, device_type: DiscreteGpu, driver: "", driver_info: "", backend: Dx12 }
[INFO  wgpu_core::instance] Adapter Dx12 AdapterInfo { name: "Microsoft Basic Render Driver", vendor: 5140, device: 140, device_type: Cpu, driver: "", driver_info: "", backend: Dx12 }
[INFO  wgpu_hal::gles::egl] Max label length: 1024
[INFO  wgpu_hal::gles::egl] Enabling GLES debug output
[INFO  wgpu_hal::gles::adapter] Vendor: Google Inc.
[INFO  wgpu_hal::gles::adapter] Renderer: ANGLE (NVIDIA GeForce RTX 3060 Laptop GPU Direct3D11 vs_5_0 ps_5_0)
[INFO  wgpu_hal::gles::adapter] Version: OpenGL ES 3.0 (ANGLE 2.1.0.57ea533f79a7)
[INFO  wgpu_hal::gles::adapter] SL version: OpenGL ES GLSL ES 3.00 (ANGLE 2.1.0.57ea533f79a7)
[INFO  wgpu_core::instance] Adapter Gl AdapterInfo { name: "ANGLE (NVIDIA GeForce RTX 3060 Laptop GPU Direct3D11 vs_5_0 ps_5_0)", vendor: 0, device: 0, device_type: Other, driver: "", driver_info: "", backend: Gl }
[INFO  iced_wgpu::window::compositor] Available adapters: [
        AdapterInfo {
            name: "NVIDIA GeForce RTX 3060 Laptop GPU",
            vendor: 4318,
            device: 9568,
            device_type: DiscreteGpu,
            driver: "NVIDIA",
            driver_info: "528.24",
            backend: Vulkan,
        },
        AdapterInfo {
            name: "NVIDIA GeForce RTX 3060 Laptop GPU",
            vendor: 4318,
            device: 9568,
            device_type: DiscreteGpu,
            driver: "",
            driver_info: "",
            backend: Dx12,
        },
        AdapterInfo {
            name: "Microsoft Basic Render Driver",
            vendor: 5140,
            device: 140,
            device_type: Cpu,
            driver: "",
            driver_info: "",
            backend: Dx12,
        },
        AdapterInfo {
            name: "ANGLE (NVIDIA GeForce RTX 3060 Laptop GPU Direct3D11 vs_5_0 ps_5_0)",
            vendor: 0,
            device: 0,
            device_type: Other,
            driver: "",
            driver_info: "",
            backend: Gl,
        },
    ]
[INFO  wgpu_hal::gles::egl] Max label length: 1024
[INFO  wgpu_hal::gles::egl] Enabling GLES debug output
[INFO  wgpu_hal::gles::adapter] Vendor: Google Inc.
[INFO  wgpu_hal::gles::adapter] Renderer: ANGLE (NVIDIA GeForce RTX 3060 Laptop GPU Direct3D11 vs_5_0 ps_5_0)
[INFO  wgpu_hal::gles::adapter] Version: OpenGL ES 3.0 (ANGLE 2.1.0.57ea533f79a7)
[INFO  wgpu_hal::gles::adapter] SL version: OpenGL ES GLSL ES 3.00 (ANGLE 2.1.0.57ea533f79a7)
[INFO  wgpu_core::instance] Adapter Vulkan AdapterInfo { name: "NVIDIA GeForce RTX 3060 Laptop GPU", vendor: 4318, device: 9568, device_type: DiscreteGpu, driver: "NVIDIA", driver_info: "528.24", backend: Vulkan }
[INFO  iced_wgpu::window::compositor] Selected: AdapterInfo {
        name: "NVIDIA GeForce RTX 3060 Laptop GPU",
        vendor: 4318,
        device: 9568,
        device_type: DiscreteGpu,
        driver: "NVIDIA",
        driver_info: "528.24",
        backend: Vulkan,
    }
[INFO  iced_wgpu::window::compositor] Selected format: Bgra8UnormSrgb
[INFO  wgpu_hal::vulkan::instance] GENERAL [Loader Message (0x0)]
        Inserted device layer VK_LAYER_OBS_HOOK (C:\ProgramData\obs-studio-hook\.\graphics-hook64.dll)
[INFO  wgpu_hal::vulkan::instance]      objects: (type: INSTANCE, hndl: 0x22b9f97b810, name: ?)
[INFO  wgpu_hal::vulkan::instance] GENERAL [Loader Message (0x0)]
        Inserted device layer VK_LAYER_NV_optimus (C:\Windows\System32\DriverStore\FileRepository\nvlti.inf_amd64_38ddcdbd562c6b02\.\nvoglv64.dll)
[INFO  wgpu_hal::vulkan::instance]      objects: (type: INSTANCE, hndl: 0x22b9f97b810, name: ?)
[INFO  wgpu_hal::vulkan::instance] GENERAL [Loader Message (0x0)]
               Using "NVIDIA GeForce RTX 3060 Laptop GPU" with driver: "C:\Windows\System32\DriverStore\FileRepository\nvlti.inf_amd64_38ddcdbd562c6b02\.\nvoglv64.dll"

[INFO  wgpu_hal::vulkan::instance]      objects: (type: INSTANCE, hndl: 0x22b9f97b810, name: ?)
[INFO  cosmic_text::font::system::std] Parsed 273 font faces in 56ms.
[INFO  wgpu_core::device] Created buffer Valid((0, 1, Vulkan)) with BufferDescriptor { label: Some("glyphon params"), size: 16, usage: BufferUsages(COPY_DST | UNIFORM), mapped_at_creation: false }
[INFO  wgpu_core::device] Created texture Valid((0, 1, Vulkan)) with TextureDescriptor { label: Some("glyphon atlas"), size: Extent3d { width: 256, height: 256, depth_or_array_layers: 1 }, mip_level_count: 1, sample_count: 1, dimension: D2, format: Rgba8UnormSrgb, usage: TextureUsages(COPY_DST | TEXTURE_BINDING), view_formats: [] }
[INFO  wgpu_core::device] Created texture Valid((1, 1, Vulkan)) with TextureDescriptor { label: Some("glyphon atlas"), size: Extent3d { width: 256, height: 256, depth_or_array_layers: 1 }, mip_level_count: 1, sample_count: 1, dimension: D2, format: R8Unorm, usage: TextureUsages(COPY_DST | TEXTURE_BINDING), view_formats: [] }
[INFO  wgpu_core::device] Created render pipeline Valid((0, 1, Vulkan)) with RenderPipelineDescriptor { label: Some("iced_wgpu::quad pipeline"), layout: Some((1, 1, Vulkan)), vertex: VertexState { stage: ProgrammableStageDescriptor { module: (1, 1, Vulkan), entry_point: "vs_main" }, buffers: [VertexBufferLayout { array_stride: 8, step_mode: Vertex, attributes: [VertexAttribute { format: Float32x2, offset: 0, shader_location: 0 }] }, VertexBufferLayout { array_stride: 68, step_mode: Instance, attributes: [VertexAttribute { format: Float32x2, offset: 0, shader_location: 1 }, VertexAttribute { format: Float32x2, offset: 8, shader_location: 2 }, VertexAttribute { format: Float32x4, offset: 16, shader_location: 3 }, VertexAttribute { format: Float32x4, offset: 32, shader_location: 4 }, VertexAttribute { format: Float32x4, offset: 48, shader_location: 5 }, VertexAttribute { format: Float32, offset: 64, shader_location: 6 }] }] }, primitive: PrimitiveState { topology: TriangleList, strip_index_format: None, front_face: Cw, cull_mode: None, unclipped_depth: false, polygon_mode: Fill, conservative: false }, depth_stencil: None, multisample: MultisampleState { count: 1, mask: 18446744073709551615, alpha_to_coverage_enabled: false }, fragment: Some(FragmentState { stage: ProgrammableStageDescriptor { module: (1, 1, Vulkan), entry_point: "fs_main" }, targets: [Some(ColorTargetState { format: Bgra8UnormSrgb, blend: Some(BlendState { color: BlendComponent { src_factor: SrcAlpha, dst_factor: OneMinusSrcAlpha, operation: Add }, alpha: BlendComponent { src_factor: One, dst_factor: OneMinusSrcAlpha, operation: Add } }), write_mask: ColorWrites(RED | GREEN | BLUE | ALPHA) })] }), multiview: None }
[INFO  wgpu_core::device] Created buffer Valid((1, 1, Vulkan)) with BufferDescriptor { label: Some("iced_wgpu::quad vertex buffer"), size: 32, usage: BufferUsages(VERTEX), mapped_at_creation: true }
[INFO  wgpu_core::device] Created buffer Valid((2, 1, Vulkan)) with BufferDescriptor { label: Some("iced_wgpu::quad index buffer"), size: 12, usage: BufferUsages(INDEX), mapped_at_creation: true }
[INFO  wgpu_core::device] Created render pipeline Valid((1, 1, Vulkan)) with RenderPipelineDescriptor { label: Some("iced_wgpu::triangle::msaa pipeline"), layout: Some((2, 1, Vulkan)), vertex: VertexState { stage: ProgrammableStageDescriptor { module: (1, 2, Vulkan), entry_point: "vs_main" }, buffers: [] }, primitive: PrimitiveState { topology: TriangleList, strip_index_format: None, front_face: Cw, cull_mode: None, unclipped_depth: false, polygon_mode: Fill, conservative: false }, depth_stencil: None, multisample: MultisampleState { count: 1, mask: 18446744073709551615, alpha_to_coverage_enabled: false }, fragment: Some(FragmentState { stage: ProgrammableStageDescriptor { module: (1, 2, Vulkan), entry_point: "fs_main" }, targets: [Some(ColorTargetState { format: Bgra8UnormSrgb, blend: Some(BlendState { color: BlendComponent { src_factor: One, dst_factor: OneMinusSrcAlpha, operation: Add }, alpha: BlendComponent { src_factor: One, dst_factor: OneMinusSrcAlpha, operation: Add } }), write_mask: ColorWrites(RED | GREEN | BLUE | ALPHA) })] }), multiview: None }
[INFO  wgpu_core::device] Created render pipeline Valid((2, 1, Vulkan)) with RenderPipelineDescriptor { label: Some("iced_wgpu::triangle::solid pipeline"), layout: Some((3, 1, Vulkan)), vertex: VertexState { stage: ProgrammableStageDescriptor { module: (1, 3, Vulkan), entry_point: "vs_main" }, buffers: [VertexBufferLayout { array_stride: 24, step_mode: Vertex, attributes: [VertexAttribute { format: Float32x2, offset: 0, shader_location: 0 }, VertexAttribute { format: Float32x4, offset: 8, shader_location: 1 }] }] }, primitive: PrimitiveState { topology: TriangleList, strip_index_format: None, front_face: Cw, cull_mode: None, unclipped_depth: false, polygon_mode: Fill, conservative: false }, depth_stencil: None, multisample: MultisampleState { count: 4, mask: 18446744073709551615, alpha_to_coverage_enabled: false }, fragment: Some(FragmentState { stage: ProgrammableStageDescriptor { module: (1, 3, Vulkan), entry_point: "fs_main" }, targets: [Some(ColorTargetState { format: Bgra8UnormSrgb, blend: Some(BlendState { color: BlendComponent { src_factor: SrcAlpha, dst_factor: OneMinusSrcAlpha, operation: Add }, alpha: BlendComponent { src_factor: One, dst_factor: OneMinusSrcAlpha, operation: Add } }), write_mask: ColorWrites(RED | GREEN | BLUE | ALPHA) })] }), multiview: None }
[INFO  wgpu_core::device] Created render pipeline Valid((3, 1, Vulkan)) with RenderPipelineDescriptor { label: Some("iced_wgpu::triangle::gradient pipeline"), layout: Some((4, 1, Vulkan)), vertex: VertexState { stage: ProgrammableStageDescriptor { module: (1, 4, Vulkan), entry_point: "vs_main" }, buffers: [VertexBufferLayout { array_stride: 8, step_mode: Vertex, attributes: [VertexAttribute { format: Float32x2, offset: 0, shader_location: 0 }] }] }, primitive: PrimitiveState { topology: TriangleList, strip_index_format: None, front_face: Cw, cull_mode: None, unclipped_depth: false, polygon_mode: Fill, conservative: false }, depth_stencil: None, multisample: MultisampleState { count: 4, mask: 18446744073709551615, alpha_to_coverage_enabled: false }, fragment: Some(FragmentState { stage: ProgrammableStageDescriptor { module: (1, 4, Vulkan), entry_point: "fs_main" }, targets: [Some(ColorTargetState { format: Bgra8UnormSrgb, blend: Some(BlendState { color: BlendComponent { src_factor: SrcAlpha, dst_factor: OneMinusSrcAlpha, operation: Add }, alpha: BlendComponent { src_factor: One, dst_factor: OneMinusSrcAlpha, operation: Add } }), write_mask: ColorWrites(RED | GREEN | BLUE | ALPHA) })] }), multiview: None }
[INFO  wgpu_core::device] configuring surface with SurfaceConfiguration { usage: TextureUsages(RENDER_ATTACHMENT), format: Bgra8UnormSrgb, width: 1280, height: 960, present_mode: AutoVsync, alpha_mode: Auto, view_formats: [] }
[INFO  wgpu_core::device] Automatically choosing presentation mode by rule AutoVsync. Chose Fifo
[INFO  wgpu_core::device] Automatically choosing alpha mode by rule Auto. Chose Opaque
[INFO  wgpu_core::device] configuring surface with SurfaceConfiguration { usage: TextureUsages(RENDER_ATTACHMENT), format: Bgra8UnormSrgb, width: 1280, height: 960, present_mode: AutoVsync, alpha_mode: Auto, view_formats: [] }
[INFO  wgpu_core::device] Automatically choosing presentation mode by rule AutoVsync. Chose Fifo
[INFO  wgpu_core::device] Automatically choosing alpha mode by rule Auto. Chose Opaque
[INFO  wgpu_core::device] Created buffer Valid((3, 1, Vulkan)) with BufferDescriptor { label: Some("iced_wgpu::quad uniforms buffer"), size: 80, usage: BufferUsages(COPY_DST | UNIFORM), mapped_at_creation: false }
[INFO  wgpu_core::device] Created buffer Valid((4, 1, Vulkan)) with BufferDescriptor { label: Some("iced_wgpu::quad instance buffer"), size: 1048576, usage: BufferUsages(COPY_DST | VERTEX), mapped_at_creation: false }
[INFO  wgpu_core::device] Created buffer Valid((5, 1, Vulkan)) with BufferDescriptor { label: Some("iced_wgpu::triangle index buffer"), size: 512, usage: BufferUsages(COPY_DST | INDEX), mapped_at_creation: false }
[INFO  wgpu_core::device] Created buffer Valid((6, 1, Vulkan)) with BufferDescriptor { label: Some("iced_wgpu::triangle::solid vertex buffer"), size: 3072, usage: BufferUsages(COPY_DST | VERTEX), mapped_at_creation: false }
[INFO  wgpu_core::device] Created buffer Valid((7, 1, Vulkan)) with BufferDescriptor { label: Some("iced_wgpu::triangle::solid uniforms"), size: 64, usage: BufferUsages(COPY_DST | UNIFORM), mapped_at_creation: false }
[INFO  wgpu_core::device] Created buffer Valid((8, 1, Vulkan)) with BufferDescriptor { label: Some("iced_wgpu::triangle::gradient vertex buffer"), size: 1024, usage: BufferUsages(COPY_DST | VERTEX), mapped_at_creation: false }
[INFO  wgpu_core::device] Created buffer Valid((9, 1, Vulkan)) with BufferDescriptor { label: Some("iced_wgpu::triangle::gradient uniforms"), size: 96, usage: BufferUsages(COPY_DST | UNIFORM), mapped_at_creation: false }
[INFO  wgpu_core::device] Created buffer Valid((10, 1, Vulkan)) with BufferDescriptor { label: Some("iced_wgpu::triangle::gradient storage"), size: 32, usage: BufferUsages(COPY_DST | STORAGE), mapped_at_creation: false }
[INFO  wgpu_core::device] Created buffer Valid((11, 1, Vulkan)) with BufferDescriptor { label: Some("iced_wgpu::triangle index buffer"), size: 4068, usage: BufferUsages(COPY_DST | INDEX), mapped_at_creation: false }
[INFO  wgpu_core::device] Created buffer Valid((12, 1, Vulkan)) with BufferDescriptor { label: Some("iced_wgpu::triangle::solid vertex buffer"), size: 16272, usage: BufferUsages(COPY_DST | VERTEX), mapped_at_creation: false }
[INFO  wgpu_core::device] Created buffer Valid((13, 1, Vulkan)) with BufferDescriptor { label: Some("iced_wgpu::triangle::solid uniforms"), size: 576, usage: BufferUsages(COPY_DST | UNIFORM), mapped_at_creation: false }
[INFO  wgpu_core::device] Created buffer Valid((14, 1, Vulkan)) with BufferDescriptor { label: Some("glyphon vertices"), size: 4096, usage: BufferUsages(COPY_DST | VERTEX), mapped_at_creation: false }
[INFO  wgpu_core::device] Created buffer Valid((15, 1, Vulkan)) with BufferDescriptor { label: Some("glyphon indices"), size: 4096, usage: BufferUsages(COPY_DST | INDEX), mapped_at_creation: false }
[INFO  wgpu_core::device] Created render pipeline Valid((4, 1, Vulkan)) with RenderPipelineDescriptor { label: Some("glyphon pipeline"), layout: Some((0, 1, Vulkan)), vertex: VertexState { stage: ProgrammableStageDescriptor { module: (0, 1, Vulkan), entry_point: "vs_main" }, buffers: [VertexBufferLayout { array_stride: 28, step_mode: Vertex, attributes: [VertexAttribute { format: Sint32x2, offset: 0, shader_location: 0 }, VertexAttribute { format: Uint32, offset: 8, shader_location: 1 }, VertexAttribute { format: Uint32, offset: 12, shader_location: 2 }, VertexAttribute { format: Uint32, offset: 16, shader_location: 3 }, VertexAttribute { format: Uint32, offset: 20, shader_location: 4 }, VertexAttribute { format: Float32, offset: 24, shader_location: 5 }] }] }, primitive: PrimitiveState { topology: TriangleList, strip_index_format: None, front_face: Ccw, cull_mode: None, unclipped_depth: false, polygon_mode: Fill, conservative: false }, depth_stencil: None, multisample: MultisampleState { count: 1, mask: 18446744073709551615, alpha_to_coverage_enabled: false }, fragment: Some(FragmentState { stage: ProgrammableStageDescriptor { module: (0, 1, Vulkan), entry_point: "fs_main" }, targets: [Some(ColorTargetState { format: Bgra8UnormSrgb, blend: Some(BlendState { color: BlendComponent { src_factor: SrcAlpha, dst_factor: OneMinusSrcAlpha, operation: Add }, alpha: BlendComponent { src_factor: One, dst_factor: OneMinusSrcAlpha, operation: Add } }), write_mask: ColorWrites(RED | GREEN | BLUE | ALPHA) })] }), multiview: None }
[INFO  wgpu_core::device] Buffer (14, 1, Vulkan) is destroyed
[INFO  wgpu_core::device] Created buffer Valid((16, 1, Vulkan)) with BufferDescriptor { label: Some("glyphon vertices"), size: 8192, usage: BufferUsages(COPY_DST | VERTEX), mapped_at_creation: true }
[INFO  wgpu_core::device] Created texture Valid((3, 1, Vulkan)) with TextureDescriptor { label: Some("iced_wgpu::triangle::msaa attachment"), size: Extent3d { width: 1280, height: 960, depth_or_array_layers: 1 }, mip_level_count: 1, sample_count: 4, dimension: D2, format: Bgra8UnormSrgb, usage: TextureUsages(RENDER_ATTACHMENT), view_formats: [] }
[INFO  wgpu_core::device] Created texture Valid((4, 1, Vulkan)) with TextureDescriptor { label: Some("iced_wgpu::triangle::msaa resolve target"), size: Extent3d { width: 1280, height: 960, depth_or_array_layers: 1 }, mip_level_count: 1, sample_count: 1, dimension: D2, format: Bgra8UnormSrgb, usage: TextureUsages(TEXTURE_BINDING | RENDER_ATTACHMENT), view_formats: [] }
[INFO  wgpu_core::hub] Dropping Global
[INFO  wgpu_core::device] Destroying 3 command encoders
[WARN  wgpu_hal::auxil::dxgi::exception] Process is terminating. Using simple reporting. Please call ReportLiveObjects() at runtime for standard reporting.
[WARN  wgpu_hal::auxil::dxgi::exception] Live Producer at 0x0000022BA5422EF0, Refcount: 2.
[WARN  wgpu_hal::auxil::dxgi::exception]        Live Object at 0x0000022BA54F8BB0, Refcount: 0.
[WARN  wgpu_hal::auxil::dxgi::exception]        Live Object at 0x0000022BA56992F0, Refcount: 0.
[WARN  wgpu_hal::auxil::dxgi::exception]        Live Object at 0x0000022BA56A7260, Refcount: 0.
[WARN  wgpu_hal::auxil::dxgi::exception]        Live Object at 0x0000022BA56B5670, Refcount: 0.
[WARN  wgpu_hal::auxil::dxgi::exception]        Live Object at 0x0000022BA56EE9C0, Refcount: 0.
[WARN  wgpu_hal::auxil::dxgi::exception]        Live Object at 0x0000022BA56EE2F0, Refcount: 0.
[WARN  wgpu_hal::auxil::dxgi::exception]        Live Object at 0x0000022BA56E6640, Refcount: 0.
[WARN  wgpu_hal::auxil::dxgi::exception]        Live Object at 0x0000022BA56E8CD0, Refcount: 0.
[WARN  wgpu_hal::auxil::dxgi::exception]        Live Object at 0x0000022BA58C1700, Refcount: 0.
[WARN  wgpu_hal::auxil::dxgi::exception] Live                         Object :      9
[WARN  wgpu_hal::auxil::dxgi::exception] Live Producer at 0x0000022BA5907B80, Refcount: 2.
[WARN  wgpu_hal::auxil::dxgi::exception]        Live Object at 0x0000022BA54F8970, Refcount: 0.
[WARN  wgpu_hal::auxil::dxgi::exception]        Live Object at 0x0000022BA593C4A0, Refcount: 0.
[WARN  wgpu_hal::auxil::dxgi::exception]        Live Object at 0x0000022BA5962410, Refcount: 0.
[WARN  wgpu_hal::auxil::dxgi::exception]        Live Object at 0x0000022BA599C530, Refcount: 0.
[WARN  wgpu_hal::auxil::dxgi::exception]        Live Object at 0x0000022BA599CCA0, Refcount: 0.
[WARN  wgpu_hal::auxil::dxgi::exception]        Live Object at 0x0000022BA5968670, Refcount: 0.
[WARN  wgpu_hal::auxil::dxgi::exception]        Live Object at 0x0000022BA58C0660, Refcount: 0.
[WARN  wgpu_hal::auxil::dxgi::exception]        Live Object at 0x0000022BA599D500, Refcount: 0.
[WARN  wgpu_hal::auxil::dxgi::exception]        Live Object at 0x0000022BA58C22E0, Refcount: 0.
[WARN  wgpu_hal::auxil::dxgi::exception] Live                         Object :      9
```
</details>

File: widget/src/pick_list.rs [hunk 1]
```diff
@@ -637,20 +684,22 @@ pub fn draw<'a, T, Renderer>(
...
             } else {
                 style.placeholder_color
             },
             bounds: Rectangle {
                 x: bounds.x + padding.left,
-                y: bounds.center_y() - text_size / 2.0,
+                y: bounds.center_y(),
                 width: bounds.width - padding.horizontal(),
-                height: text_size,
+                height: text_size * 1.2,
```
Comment (tarkah) [inline widget/src/pick_list.rs:698]: Hard-coded `1.2` line height value is used here when we have `text_line_height` that should be used

**Comment by tarkah** [comment]
> image
> 
> I just noticed after testing the tour example that any longish glyph (like the `g`s here) is getting cut off at the bottom (tested on windows & linux). Adjusting the `to_absolute` function for the logical pixels of `LineHeight` and just adding some manual padding makes the long characters render correctly, so I'm assuming the issue is with the new `LineHeight` changes. Or perhaps longer glyphs go outside the "bounds" of the text?

I'm noticing the same issue

**Comment by hecrj** [comment]
@bungoboingo @tarkah This is caused by 9a8b30d7e9cf76c097d68c84736e7687b452d5c0.

I have incremented the default line height to `1.3` since it seems more readable to me and should prevent this issue. It may be interesting to eventually expose a flag to control the clipping behavior on overflow.

File: widget/src/pick_list.rs (hunk 1 above)
Comment (hecrj) [reply widget/src/pick_list.rs:709]: Fixed in f0c87375d53d620b5939d87e1bc54288f9c184ac.

Comment (tarkah) [reply widget/src/pick_list.rs:709]: @hecrj Just a heads up, this only addressed the icon height, but not the height / bounds of the actual text (last fill_text of the draw helper)

Comment (hecrj) [reply widget/src/pick_list.rs]: :facepalm: Fixed in 1400b5187d6fa30f5fa4f83684939f6dc87cb55f.

**Comment by hecrj** [comment]
@bungoboingo I would expect the same. Seems it's an issue others have encountered as well: https://github.com/pop-os/cosmic-text/issues/123.

**Comment by hecrj** [comment]
Hopefully, the `1.3` default line height makes the issue less common. If it still persists, I can revert 9a8b30d7e9cf76c097d68c84736e7687b452d5c0 for now.

**Comment by i509VCB** [comment]
One question I have is whether the candidate backend selection should order backends based on the most performant option. I'd think something like this:
1. All hardware adapters for wgpu in the order wgpu enumerates
2. Software renderer
3. Software vulkan implementations (via wgpu)

https://github.com/iced-rs/iced/blob/advanced-text/renderer/src/compositor.rs#L155

**Comment by i509VCB** [comment]
> Can we reliably guarantee that our software rendering implementation is more performant than the driver virtual implementations? I don't know if this has an obvious answer or we need to run some actual numbers.

I guess the answer is it depends. Lavapipe might be able to use AVX-512 if supported but then there is the issue of compiler overhead. tiny-skia does have some simd support.

Probably needs to be measured anyways.

**Comment by hecrj** [comment]
@gabrieldechichi I was able to reproduce and fix the issue in #1847.

**Pull request merged**